Converted inline scripts to external scripts.

alanverresen opened the pull request:
**WORK IN PROGRESS: DO NOT MERGE WITHOUT FURTHER DISCUSSION (HERE OR #101)**

I've tried to convert the inline scripts used by widgets to external scripts without making any changes that would break any existing projects. I was a bit overzealous by removing the method that injects the `action` variable into the V3 widget, because that would break custom widget templates that use that variable, hence the third commit.

This does introduce one change that is potentially breaking: developers need to define a STATIC_ROOT to use static files. In my mind, (almost) every project would set this setting, but it's a breaking change nonetheless. Anyone got any insight in this?

I had to remove the test assertions related to the inline scripts, because those do not exist anymore. I think actual e2e tests should be used in the future to make sure that the implementation actually works.

I tried to replicate the approach/behavior of the original inline scripts in the new external scripts in the same way as the original inline scripts, so that you can compare the two, but there are some major differences:

- I wrapped every script in an IIFE to avoid namespace pollution [1]
- the execution of the scripts is delayed until after the document has been loaded
- I used guards for more feedback; they also prevent breaking things when multiple reCAPTCHA fields are on one page, and/or (at least) one of those reCAPTCHA fields wasn't added by django_recaptcha
- in contrast to injecting variables into the inline scripts of widget templates, the external scripts now check what the name of the callback function should be by looking at the `data-callback` attribute, and then adding the callback function to the global namespace with that name
- similarly, the script of the V3 widget now checks what the action is by looking at the `data-action` attribute

One design choice that I need help with is how the static resource of each widget is defined. Which seems better?

- current solution: injecting the URL into the template 
- normal solution: defining a Media class for the Widget (see [2]) 

I tested these changes against a django project that I whipped up quickly to test each widget and "they worked on my machine". If anyone has any questions, I'll gladly answer them.

[1] [IIFE - MDN Web Docs Glossary](https://developer.mozilla.org/en-US/docs/Glossary/IIFE)
[2] [Form Assets | Django documentation](https://docs.djangoproject.com/en/4.2/topics/forms/media/#assets-as-a-static-definition)

alanverresen committed:
Removed test assertions related to inline scripts in widgets.

alanverresen committed:
Converted inline scripts to external scripts.

alanverresen committed:
Readded code to prevent breaking custom templates.

Stormheg commented:
Awesome work @alanverresen 👍👍

I'll check this in detail later! From a glance over, it looks great already!

Quick question: did you happen to test with a strict CSP policy?

Stormheg commented on django_recaptcha/static/django_recaptcha/js/widget_v3.js:
```diff
@@ -0,0 +1,50 @@
+          break;
+        }
+
+        const action = {};
+        Object.assign(action, {action: captchaElement.getAttribute("data-action")});
+
+        formElement.addEventListener("submit", function (event) {
+          event.preventDefault();
+          grecaptcha.execute(publicKey, action)
```
**Suggestion:** make it more clear what this variable is used for and avoid potential confusion

```suggestion
        const config = {};
        Object.assign(config, {action: captchaElement.getAttribute("data-action")});

        formElement.addEventListener("submit", function (event) {
          event.preventDefault();
          grecaptcha.execute(publicKey, config)
```

alanverresen replied:
I actually rewrote this part several times before, but eventually decided on using the variable name matching [1]. Using `Object.assign()` also isn't the most clear, so I changed that as well in my new commit, with an additional comment to explain the situation.

Stormheg commented on django_recaptcha/static/django_recaptcha/js/widget_v3.js:
```diff
@@ -0,0 +1,50 @@
+
+        const formElement = captchaElement.closest("form");
+        if (!formElement) {
+          console.warn("reCAPTCHA widget with UUID '" + widgetUuid + "' is not part of a form");
+          break;
+        }
+
+        const publicKey = captchaElement.getAttribute("data-sitekey");
+        if (publicKey === null) {
+          console.warn("public key missing missing from reCAPTCHA widget with UUID '" + widgetUuid + "'");
+          break;
```
**Suggestion:** please use template literals for string interpolation for consistency with the other scripts.

Stormheg commented on django_recaptcha/static/django_recaptcha/js/widget_v3.js:
```diff
@@ -0,0 +1,50 @@
+          event.preventDefault();
+          grecaptcha.execute(publicKey, action)
+            .then(function (token) {
+              console.log("reCAPTCHA validated for reCAPTCHA widget with UUID '" + widgetUuid + "'");
```
Template literal

Stormheg reviewed (commented):
Again, thank you so much for taking on the effort of converting all inline script usage! I've completed a detailed review of this PR.

Some minor comments on the code.

## Testing

I tested this using my own sandbox project with django-csp installed and in report mode and did not see any CSP violation errors. Yay! 🥇 

django-csp config used:

- `CSP_SCRIPT_SRC="'self',https://www.google.com/recaptcha/,https://www.gstatic.com/recaptcha/"`
- `CSP_FRAME_SRC="https://www.google.com/recaptcha/,https://recaptcha.google.com/recaptcha/"`


## In response to your comments

1. 
> This does introduce one change that is potentially breaking: developers need to define a STATIC_ROOT to use static files. In my mind, (almost) every project would set this setting, but it's a breaking change nonetheless. Anyone got any insight in this?

I wouldn't consider this a breaking change. Virtually all packages depend on Django's staticfiles framework being used. I deem it very likely people will already have this set up and configured. In other words: this is unlikely to affect people.

2. 
> I had to remove the test assertions related to the inline scripts, because those do not exist anymore. I think actual e2e tests should be used in the future to make sure that the implementation actually works.

Yes, yes and yes to E2E tests! A test suite that opens a browser, loads a form, and verifies there are no JS errors / CSP violations sounds like a generally good idea to me. There is some maintenance overhead in keeping this up to date with current browsers, but I believe this package has enough momentum that that shouldn't be too much of an issue.

Given we currently have no e2e test suite I would consider it out of scope to add it in this PR. Good candidate for a follow up PR/issue if you are up for it.

3. 
> One design choice that I need help with is how the static resource of each widget is defined. Which seems better?
>
> - current solution: injecting the URL into the template
> - normal solution: defining a Media class for the Widget (see [2])

Defining a media class would be a breaking change I think. Form media isn't always rendered, depending on how the form is used in a template. [There has also been discussion about potentially deprecating form media](https://forum.djangoproject.com/t/rejuvenating-vs-deprecating-form-media/21285/11). All this together makes me hesitate in recommending this approach.

## Static file url

As for passing the static file url to the template, why can't we just do this in the template?

```dj+html
{% load static %}
<script src="{% static "django_recaptcha/js/widget_v3.js" %}></script>
```

The current solution that passes the static file url to template would allow the JS to be more easily customized, I think?

But I also suspect such customizations would be very fragile in practice so we shouldn't be encouraging this too much by allowing easy customisation. Those desperate enough can override the template and by doing so accept responsibility in keeping their customisations compatible with new releases.

## Potential caveat

A potential caveat to keep in mind is static file caching. Browsers/CDNs might cache the JS and when django recaptcha is updated, the old JS code might still end up being used by the client. We could consider this the problem of the developer – they haven't implemented a proper cache busting mechanism, by using something like [Whitenoise](https://whitenoise.readthedocs.io/en/latest/) for example.

We could also do something similar to what Wagtail did in an effort to combat this issue: [versioned_static](https://github.com/wagtail/wagtail/blob/08800787758c1175e82f27d0aaaecd2429b6db39/wagtail/admin/staticfiles.py#L47).

I guess it depends on how often we the JS is going to change dramatically, which I assume is going to be very rare.

alanverresen committed:
Changed string concatenations to string literals.

alanverresen committed:
Rewrote script based on feedback.

alanverresen committed:
Use static template tag to include scripts.

alanverresen commented:
> I wouldn't consider this a breaking change. Virtually all packages depend on Django's staticfiles framework being used. I deem it very likely people will already have this set up and configured. In other words: this is unlikely to affect people.

Yes, I tend to agree.

---

> Given we currently have no e2e test suite I would consider it out of scope to add it in this PR. Good candidate for a follow up PR/issue if you are up for it.

I've seen the new issue and will try to contribute.

---

> Defining a media class would be breaking change I think. Form media isn't always rendered, depending on how the form is used in a template. [There has also been discussion about potentially deprecating form media](https://forum.djangoproject.com/t/rejuvenating-vs-deprecating-form-media/21285/11). All this together makes me hesititate in recommending this approach.

I wasn't aware of this, but then the Media class definitely isn't a good option. I went with the solution that you presented, which is much more in line with common practices.

I expect that developers will understand that if it's their own responsibility if they override the template.

---

> A potential caveat to keep in mind is static file caching. Browsers/CDNs might cache the JS and when django recaptcha is updated, the old JS code might still end up being used by the client. We could consider this the problem of the developer – they haven't implemented a proper cache busting mechanism, by using something like [Whitenoise](https://whitenoise.readthedocs.io/en/latest/) for example.
> 
> We could also do something similar to what Wagtail did in an effort to combat this issue: [versioned_static](https://github.com/wagtail/wagtail/blob/08800787758c1175e82f27d0aaaecd2429b6db39/wagtail/admin/staticfiles.py#L47).
> 
> I guess it depends on how often we the JS is going to change dramatically, which I assume is going to be very rare.

I hadn't considered caching problems yet, but this is indeed a problem now that we're changing the inline scripts to external scripts. I think that introducing a solution like the one used by Wagtail would ensure that we're never going to break someone that upgrades their package.

Stormheg commented on django_recaptcha/static/django_recaptcha/js/widget_v3.js:
```diff
@@ -0,0 +1,52 @@
+          break;
+        }
+
+        // Google reCAPTCHA expects an object with an optional single key-value
+        // pair to communicate the name of the action.
+        const actionName = captchaElement.getAttribute("data-action");
+        const config = actionName !== null ? { action: actionName } : {};
```
**Nitpick:** @alanverresen I don't think this rewritten code makes things much clearer. What about this?

 ```suggestion
         const actionName = captchaElement.getAttribute("data-action");
         const config = {};
         if (actionName !== null) {
           config.action = actionName;
         }
```

alanverresen replied:
I'm fine with both, so if you think yours is clearer, we'll go with that.

Stormheg commented:
> I hadn't considered caching problems yet, but this is indeed a problem now that we're changing the inline scripts to external scripts. I think that introducing a solution like the one used by Wagtail would ensure that we're never going to break someone that upgrades their package.

I'm somewhat hesitant to adopting a custom solution like Wagtail (even though I suggested it as an approach). Yes, it ensures people never run into caching issues, which means we hopefully don't have to deal with bug reports reporting strange issues caused by caching.

But how often will we change the JS in a backwards incompatible way? Not super often I think – the JS should be pretty stable. Is this argument really enough to justify doing something custom for cache busting? I hope that most people have a cache busting mechanism set up for their static files so a custom solution on our end shouldn't be necessary.

Could use more opinions on this, calling @django-recaptcha/contributors for insights 🧠

alanverresen committed:
Update django_recaptcha/static/django_recaptcha/js/widget_v3.js

Co-authored-by: Storm Heg <storm@stormheg.co>

alanverresen commented:
I think the safest option is to not implement a cache busting solution for now, and to add it if it ever becomes clear in the future that we actually need it.  Are there any more aspects that need to be addressed before this pull request can be merged? There's probably room for improvement, but the main issue has been addressed.

thibaudcolas commented on django_recaptcha/static/django_recaptcha/js/widget_v2_invisible.js:
```diff
@@ -0,0 +1,49 @@
+        break;
+      }
+
+      const formElement = captchaElement.closest("form");
+      if (!formElement) {
```
I believe this could just be:

```suggestion
      if (! captchaElement.form) {
```

thibaudcolas commented on django_recaptcha/static/django_recaptcha/js/widget_v2_checkbox.js:
```diff
@@ -0,0 +1,36 @@
+  function main() {
+    for (const captchaElement of document.querySelectorAll(".g-recaptcha")) {
+
+      const widgetUuid = captchaElement.getAttribute("data-widget-uuid");
```
```suggestion
      const widgetUUID = captchaElement.getAttribute("data-widget-uuid");
```

Should use the acronym’s case

thibaudcolas commented on django_recaptcha/static/django_recaptcha/js/widget_v2_checkbox.js:
```diff
@@ -0,0 +1,36 @@
+        console.warn(`callback function missing from reCAPTCHA widget with UUID '${widgetUuid}'`);
+        break;
+      }
+      if (window.hasOwnProperty(callbackFunctionName)) {
```
```suggestion
      if (window[callbackFunctionName]) {
```

I think it’s simpler and clearer if we access the same way as we set.

thibaudcolas commented on django_recaptcha/static/django_recaptcha/js/widget_v2_checkbox.js:
```diff
@@ -0,0 +1,36 @@
+      console.log(`found reCAPTCHA widget with UUID '${widgetUuid}'`);
+
+      const callbackFunctionName = captchaElement.getAttribute("data-callback");
+      if (!callbackFunctionName) {
+        console.warn(`callback function missing from reCAPTCHA widget with UUID '${widgetUuid}'`);
+        break;
+      }
```
It’s unclear to me when / if ever this would happen, considering the attribute is set by the project.

If there was a risk of this being unset, feels like we’d be better of with validation server-side so this shows in server logs?

alanverresen replied:
I removed the warning, but I opted to keep the guard statement as it protects the script's integrity. Not a single potential situation should crash the whole script, how unlikely it may be.

thibaudcolas commented on django_recaptcha/static/django_recaptcha/js/widget_v2_checkbox.js:
```diff
@@ -0,0 +1,36 @@
+    for (const captchaElement of document.querySelectorAll(".g-recaptcha")) {
+
+      const widgetUuid = captchaElement.getAttribute("data-widget-uuid");
+      if (!widgetUuid) {
+        console.warn("reCAPTCHA widget with missing UUID");
+        break;
+      }
```
I don’t see in what scenario this could happen. If it’s possible, would recommend adding server logs for this.

thibaudcolas commented on django_recaptcha/static/django_recaptcha/js/widget_v2_checkbox.js:
```diff
@@ -0,0 +1,36 @@
+        console.warn("reCAPTCHA widget with missing UUID");
+        break;
+      }
+      console.log(`found reCAPTCHA widget with UUID '${widgetUuid}'`);
```
This strikes me as too verbose, we shouldn’t be adding extra logging to projects that use those widgets.

thibaudcolas commented on django_recaptcha/static/django_recaptcha/js/widget_v2_checkbox.js:
```diff
@@ -0,0 +1,36 @@
+(function () {
```
This wouldn’t be needed if we used ES modules instead.

thibaudcolas commented on django_recaptcha/static/django_recaptcha/js/widget_v2_invisible.js:
```diff
@@ -0,0 +1,49 @@
+      });
+
+      window[callbackFunctionName] = (token) => {
+        console.log(`reCAPTCHA validated for reCAPTCHA widget with UUID '${widgetUuid}'`);
+        console.log("submitting form...");
```
For the purpose of playing well with existing error monitoring, I’d recommend using the same number of console log statements as before unless it really helps to add more separate ones.

alanverresen replied:
I removed all console logs, and only kept warnings for situations where they were actually helpful.

thibaudcolas commented on django_recaptcha/static/django_recaptcha/js/widget_v3.js:
```diff
@@ -0,0 +1,53 @@
+        }
+
+        const publicKey = captchaElement.getAttribute("data-sitekey");
+        if (publicKey === null) {
+          console.warn(`public key missing missing from reCAPTCHA widget with UUID '${widgetUuid}'`);
+          break;
+        }
```
I don’t see in what scenario this is possible, it looks like we always have a value for this.

thibaudcolas reviewed (commented):
With the caveat that I don’t have much experience with reCAPTCHA myself,

- The new code looks like it would be incompatible with one page using different widgets ? The initilisation selector would need to be different per widget type at least.
- There’s lots more defensive coding here which I think would be better placed in Python _if_ needed at all

As far as how to load the JS, I don’t see a use case for having two reCAPTCHA fields on one form, so I don’t see the usefulness of `Media`. We do need to account for multiple forms on the one page using reCAPTCHA, but Media doesn’t help with that.

I agree an end-to-end test suite would be great. I’d recommend starting just with a demo site in a `demo` folder or similar, so people like me who don’t work with this often can have something up and running quickly even if there were no automated tests.

Stormheg commented:
Hey @alanverresen - are you still interested in working on this? If availability it is an issue, I could probably find some time to help push this over the finish line.

alanverresen commented:
@Stormheg Sorry for the late answer. I'll take a look at this again this week, and also incorporate the helpful feedback provided by @thibaudcolas (thanks btw!). I don't know how much work it'll take to bring this back up-to-date with the main branch; perhaps I'll create a new pull request based on the current main branch, instead.

alanverresen commented:
Feel free to do it yourself if you're certain that it won't take too much effort or time!

Stormheg commented:
@alanverresen done, I pushed to your fork.

Since rebasing alters git history, you should reset your local branch to the version that is on your fork. The easiest way to do that  -- assuming your remote is called `origin` -- `git reset --hard origin/issue101-externalscripts    `

alanverresen commented:
@Stormheg Awesome, thanks! I think I've got everything set up correctly to continue working on this.

I went through all past commits and comments again.
I will definitely be making the following changes based on the feedback by @thibaudcolas:

- switching from IIFE to ES modules
- use captchaElement.form
- use the name widgetUUID instead of widgetUuid
- use [] instead of .hasOwnProperty
- remove unnecessary checks
- remove console logs or replace them with warnings when there's an actual issue

There's also the other issue that was brought up by @thibaudcolas: this solution cannot handle multiple instances of different widgets on one page. I will handle this by also injecting a data attribute into each widget, so that scripts can distinguish between different types of widgets. This way, a script won't add an event listener to reCAPTCHA widgets of the wrong type.

In addition, I will also remove the comment found in every widget's template that the implementation only caters for that single widget, because this isn't really true anymore. For example, if you add two reCAPTCHA widgets of the same type to a single page, then two copies of the same script will be executed: one of them will handle adding callback functions and event listeners (or equivalent) for both widgets, the other won't do anything.

alanverresen committed:
Switched from IIFE to ES modules for widget scripts.

alanverresen committed:
Use the name widgetUUID instead of widgetUuid.

alanverresen committed:
Use square bracket notation to dynamically access window property.

alanverresen committed:
Replaced break statements with continue statements.

An issue with one reCAPTCHA widget should not affect other widgets.
Hence, continue statements should be used in break statements.

alanverresen committed:
Removed console logs.

alanverresen committed:
Added data attribute to type of reCAPTCHA.

alanverresen committed:
Removed outdated comment in widget templates.

This new implementation should be able to handle multiple reCAPTCHAs of
the same or different types on the same page. It also shouldn't clash
with reCAPTCHAs added by other libraries.

Each widget will inject one copy of the script to set up reCAPTCHAs of
its respective type. As as a result, multiple copies of the same script
may be injected, but only one of these scripts will set up all the
reCAPTCHAs of the right type that were added by django-recaptcha. The
script's guards will prevent other copies from doing anything.

alanverresen commented:
I have now made all the changes that I listed in https://github.com/django-recaptcha/django-recaptcha/pull/336#issuecomment-2864270573, with some notes:

- using captchaElement.form was not possible, as discussed in https://github.com/django-recaptcha/django-recaptcha/pull/336#discussion_r2083468150
- I chose the following names to show a contrast between the "old" reCAPTCHA and reCAPTCHA enterprise, for when it is added:
  * classic-v2-checkbox
  * classic-v2-invisible
  * classic-v3
  * enterprise-v1

I do not have access to classic v3 keys anymore, so I couldn't test my solution for reCAPTCHA v3. However, because the changes in these commits barely touched any logic, I'm fairly certain that v3 should still work. Can anyone confirm that it still works, or privately share a pair of test keys that I can use to test v3 locally.

alanverresen commented:
I'm still not satisfied with the current state of this solution. I've been taking a deeper look into how reCAPTCHA should be integrated on the client side, and still need to think some things through before I can confidently say what's okay and what needs to change.

Stormheg commented:
Thanks @alanverresen - I'll put this on my review backlog.

Stormheg commented:
@alanverresen this has been on my review backlog a long time, sorry about that... I'll likely get to it during the quiet holiday season 🤞

mgax commented on django_recaptcha/static/django_recaptcha/js/widget_v3.js:
```diff
@@ -0,0 +1,50 @@
+
+      const publicKey = captchaElement.getAttribute("data-sitekey");
+      if (publicKey === null) {
+        console.warn(`public key missing missing from reCAPTCHA widget with UUID '${widgetUUID}'`);
```
Small typo:

```suggestion
        console.warn(`public key missing from reCAPTCHA widget with UUID '${widgetUUID}'`);
```

mgax commented on django_recaptcha/static/django_recaptcha/js/widget_v2_invisible.js:
```diff
@@ -0,0 +1,45 @@
+
+    formElement.addEventListener("submit", (event) => {
+      event.preventDefault();
+      grecaptcha.execute();
```
This call is not specific to our `captchaElement` here. I believe it triggers the _first_ captcha on the page. This would mean that, if there are 2 forms on the page, each with its own captcha, the user submitting the 2nd form ends up submitting the 1st form instead.

I think the easiest solution is to restore the comment in `django_recaptcha/templates/django_recaptcha/includes/js_v2_invisible.html` that said "The provided implementation caters for only one reCAPTCHA on a page."

mgax reviewed (commented):
Hi @alanverresen, thank you for persisting with this PR, I think it is in pretty good shape already!

I would suggest two additions:
- Update the changelog and give a heads up that any project that overrides the templates should review them.
- Add a short one-line mention in the Readme's "Installation" section to make sure static files are configured, with a link to https://docs.djangoproject.com/en/stable/ref/contrib/staticfiles/.
